src/array.rs

```diff
@@ -0,0 +1,456 @@
+        assert_eq!(array.item::<bool>(), true);
+        assert_eq!(array.itemsize(), 1);
+        assert_eq!(array.size(), 1);
+        assert!(array.strides().is_empty());
+        assert_eq!(array.nbytes(), 1);
+        assert_eq!(array.ndim(), 0);
+        assert!(array.shape().is_empty());
+        assert_eq!(array.dtype(), Dtype::Bool);
+    }
+
+    #[test]
+    fn new_scalar_array_from_int() {
+        let array = Array::from_int(42);
+        assert_eq!(array.item::<i32>(), 42);
+        assert_eq!(array.itemsize(), 4);
+        assert_eq!(array.size(), 1);
+        assert!(array.strides().is_empty());
+        assert_eq!(array.nbytes(), 4);
+        assert_eq!(array.ndim(), 0);
+        assert!(array.shape().is_empty());
+        assert_eq!(array.dtype(), Dtype::Int32);
+    }
+
+    #[test]
+    fn new_scalar_array_from_float() {
+        let array = Array::from_float(3.14);
+        assert_eq!(array.item::<f32>(), 3.14);
+        assert_eq!(array.itemsize(), 4);
+        assert_eq!(array.size(), 1);
+        assert!(array.strides().is_empty());
+        assert_eq!(array.nbytes(), 4);
+        assert_eq!(array.ndim(), 0);
+        assert!(array.shape().is_empty());
+        assert_eq!(array.dtype(), Dtype::Float32);
+    }
+
+    #[test]
+    fn new_scalar_array_from_complex() {
+        let val = c64 { re: 1.0, im: 2.0 };
+        let array = Array::from_complex(val);
+        assert_eq!(array.item::<c64>(), val);
+        assert_eq!(array.itemsize(), 8);
+        assert_eq!(array.size(), 1);
+        assert!(array.strides().is_empty());
+        assert_eq!(array.nbytes(), 8);
+        assert_eq!(array.ndim(), 0);
+        assert!(array.shape().is_empty());
+        assert_eq!(array.dtype(), Dtype::Complex64);
+    }
+
+    #[test]
+    fn new_array_from_single_element_slice() {
+        let data = [1i32];
+        let array = Array::from_data(&data, &[1], 1);
+        assert_eq!(array.data::<i32>(), &data);
+        assert_eq!(array.item::<i32>(), 1);
+        assert_eq!(array.itemsize(), 4);
+        assert_eq!(array.size(), 1);
+        assert_eq!(array.strides(), &[1]);
+        assert_eq!(array.nbytes(), 4);
+        assert_eq!(array.ndim(), 1);
+        assert_eq!(array.dim(0), 1);
+        assert_eq!(array.shape(), &[1]);
+        assert_eq!(array.dtype(), Dtype::Int32);
+    }
+
+    #[test]
+    fn new_array_from_multi_element_slice() {
+        let data = [1i32, 2, 3, 4, 5];
+        let array = Array::from_data(&data, &[5], 1);
+        assert_eq!(array.data::<i32>(), &data);
+        assert_eq!(array.itemsize(), 4);
+        assert_eq!(array.size(), 5);
+        assert_eq!(array.strides(), &[1]);
+        assert_eq!(array.nbytes(), 20);
+        assert_eq!(array.ndim(), 1);
+        assert_eq!(array.dim(0), 5);
+        assert_eq!(array.shape(), &[5]);
+        assert_eq!(array.dtype(), Dtype::Int32);
+    }
+
+    // // TODO: fatal runtime error: Rust cannot catch foreign exceptions
```

Comment:
I think if we start doing the precondition check as I mentioned in another comment, we should no longer get these issues I believe.

Reply:
The c++ exceptions are quite nested though, so might be hard to cover all of them. But I agree that this would probably get rid of the problem with c++ exceptions

Reply:
Perform checks in rust to avoid c++ exception